core-bundle/src/EventListener/DataContainer/Undo/LabelListener.php

```diff
@@ -0,0 +1,195 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer\Undo;
+
+use Contao\ArrayUtil;
+use Contao\Config;
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\ServiceAnnotation\Callback;
+use Contao\DataContainer;
+use Contao\Date;
+use Contao\StringUtil;
+use Contao\System;
+use Contao\UserModel;
+use Doctrine\DBAL\Connection;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use Twig\Environment;
+
+/**
+ * @Callback(target="list.label.label", table="tl_undo")
+ *
+ * @internal
+ */
+class LabelListener
+{
+    use UndoListenerTrait;
+
+    private Environment $twig;
+
+    public function __construct(ContaoFramework $framework, Connection $connection, TranslatorInterface $translator, Environment $twig)
+    {
+        $this->framework = $framework;
+        $this->connection = $connection;
+        $this->translator = $translator;
+        $this->twig = $twig;
+    }
+
+    public function __invoke(array $row, string $label, DataContainer $dc): string
+    {
+        $this->framework->initialize();
+
+        $table = $row['fromTable'];
+        $originalRow = StringUtil::deserialize($row['data'])[$table][0];
+
+        $controller = $this->framework->getAdapter(Controller::class);
+        $controller->loadDataContainer($table);
+
+        return $this->twig->render(
+            '@ContaoCore/Backend/be_undo_label.html.twig',
+            $this->getTemplateData($table, $row, $originalRow)
+        );
+    }
+
+    private function getTemplateData(string $table, array $row, array $originalRow): array
+    {
+        $dataContainer = $this->framework->getAdapter(DataContainer::class)->getDriverForTable($table);
+        $originalTableDc = new $dataContainer($table);
+        $user = $this->framework->getAdapter(UserModel::class)->findById($row['pid']);
+        $config = $this->framework->getAdapter(Config::class);
+        $parent = null;
+
+        if (true === ($GLOBALS['TL_DCA'][$table]['config']['dynamicPtable'] ?? null)) {
+            $parent = ['table' => $originalRow['ptable'], 'id' => $originalRow['pid']];
+        }
+
+        return [
+            'preview' => $this->renderPreview($originalRow, $originalTableDc),
+            'user' => $user,
+            'row' => $row,
+            'translatedFromTable' => $this->getTranslatedTypeFromTable($table),
```

Comment:
Please apply translations in the template using `|trans`.

Reply:
AFAIK this is not possible because I cannot define a fallback in the template. If a table is not translated (e.g. an extension is not yet updated), I want to display the table name as fallback, but the Twig filter would use the translation key as fallback (e.g. `TABLES.tl_content.0` rather than `tl_content`)

Reply:
We probably want to get rid of the translation anyway 🙃

Reply:
> AFAIK this is not possible

Why? You could use the same fall back in the template as you currently did in `getTranslatedTypeFromTable()`. But getting rid of them would of course solve this as well. :smile:

Reply:
I removed the translations.